cupy/fft/fft.py

```diff
@@ -356,8 +356,8 @@ def _exec_fftn(a, direction, value_type, norm, axes, overwrite_x,
     if fft_type not in [cufft.CUFFT_C2C, cufft.CUFFT_Z2Z]:
         raise NotImplementedError('Only C2C and Z2Z are supported.')
 
-    if a.base is not None:
-        a = a.copy()
+    if a.base is not None and not a.flags.forc:
+        a = a.copy(order='A')
```

Comment:
The change to the line is essential because the default `order='K'` may leave `a` non-contiguous.

(Optional) Could you also test non-contiguous view to detect it?

Reply:
I added a `test_noncontiguous_view`. Please see if that addresses the case you were thinking of

Reply:
It seems `a.copy(...)` is just verbose.

- Only `_fftn` calls `_exec_fftn` and C- or F-contiguous `a` seems passed.
- In my environment all the tests pass without these two lines.

Reply:
> In my environment all the tests pass without these two lines.

You mean ALL of the fftpack tests or just the newly added tests? For me the newly added tests pass without those two lines, but many other fftpack tests fail.

Specifically, running `pytest tests/cupyx_tests/scipy_tests/fftpack_tests/test_fftpack.py` gives `152 failed, 653 passed `. Similarly for the `cupy.fft` and `cupyx.scipy.fft` tests.

All cases pass for me with those two lines present.

Reply:
In my environment below, `pytest tests/cupyx_tests/scipy_tests/fftpack_tests/test_fftpack.py` passed without the lines on `pull/3034/merge`.  With `assert a.flags.forc`, the result was the same .

CUDA 9.0 may be too old.  With `assert False` here, the run gave `9 failed, 796 passed`.  Only 3 existing tests (in `TestFft2_param_0_{axes=None, s=None, shape=(3, 4)}`) failed.

```
CuPy Version          : 8.0.0b1
CUDA Root             : /usr/local/cuda
CUDA Build Version    : 9000
CUDA Driver Version   : 9000
CUDA Runtime Version  : 9000
cuBLAS Version        : 9000
cuFFT Version         : 9000
cuRAND Version        : 9000
cuSOLVER Version      : (9, 0, 0)
cuSPARSE Version      : 9000
NVRTC Version         : (9, 0)
cuDNN Build Version   : 7300
cuDNN Version         : 7300
NCCL Build Version    : 2005
NCCL Runtime Version  : (unknown)
```

Reply:
I reproduced `152 failed, 653 passed` after removing `a = a.copy()` in **`_exec_fft`**, where the copy is necessary (possibly because of the late `swapaxes` in the current implementation).

Reply:
Okay, I did a clean rebuild of the branch and now all tests are passing without those two lines. Apologies for the confusion!

Indeed `_fftn` always enforces contiguity so it does not necessarily need to also be checked with `_exec_fftn`.

I now get all fft-related tests passing (aside from a few known irfftn failures due to a CUFFT bug in CUDA 10.2).

Reply:
Oh sorry I failed to recognize this when reviewing! 22cede9 is a much better fix.